src/processor/call_handler.rs

```diff
@@ -0,0 +1,130 @@
+use crate::args::CallHandlerArgs;
+use crate::consts::EXTERNAL_CALL_HANDLER_DISCRIMINATOR;
+use crate::ephemeral_balance_seeds_from_payer;
+use crate::processor::utils::loaders::{
+    load_initialized_validator_fees_vault, load_owned_pda, load_pda, load_signer,
+};
+
+use borsh::BorshDeserialize;
+use solana_program::account_info::next_account_info;
+use solana_program::account_info::AccountInfo;
+use solana_program::entrypoint::ProgramResult;
+use solana_program::instruction::{AccountMeta, Instruction};
+use solana_program::program::invoke_signed;
+use solana_program::program_error::ProgramError;
+use solana_program::pubkey::Pubkey;
+use solana_program::{msg, system_program};
+
+pub const INVALID_ESCROW_PDA: &str = "invalid escrow pda in CallHandler";
+pub const INVALID_ESCROW_OWNER: &str = "escrow can not be delegated in CallHandler";
+pub const ACTION_FAILED_MSG: &str = "user action failed";
+
+/// Delegates an account
+///
+/// Accounts:
+/// 0: `[signer]`   validator
+/// 1: `[]`         validator fee vault to verify its registration
+/// 2: `[]`         destination program of an action
+/// 3: `[]`         escrow authority account which created escrow account
+/// 4: `[writable]` non delegated escrow pda created from 3
+/// 5: `[readonly/writable]` other accounts needed for action
+/// 6: `[readonly/writable]` other accounts needed for action
+/// 7: ...
+///
+/// Requirements:
+///
+/// - escrow account initialized
+/// - escrow account not delegated
+/// - validator as a caller
+///
+/// Steps:
+/// 1. Verify that signer is a valid registered validator
+/// 2. Verify escrow pda exists and not delegated
+/// 3. Invoke signed on behalf of escrow pda user specified action
+///
+/// Usage:
+///
+/// This instruction is meant to be called via CPI with the owning program signing for the
+/// delegated account.
+pub fn process_call_handler(
+    _program_id: &Pubkey,
+    accounts: &[AccountInfo],
+    data: &[u8],
+) -> ProgramResult {
+    let args = CallHandlerArgs::try_from_slice(data)?;
+
+    let accounts_iter = &mut accounts.iter();
+
+    let validator = next_account_info(accounts_iter)?;
+    let validator_fees_vault = next_account_info(accounts_iter)?;
+    let destination_program = next_account_info(accounts_iter)?;
+    let escrow_authority_account = next_account_info(accounts_iter)?;
+    let escrow_account = next_account_info(accounts_iter)?;
+
+    // verify account is a signer
+    load_signer(validator, "validator")?;
+    // verify signer is a registered validator
+    load_initialized_validator_fees_vault(validator, validator_fees_vault, true)?;
+    // Check if destination program is executable
+    if !destination_program.executable {
+        msg!(
+            "{} program is not executable: destination program",
+            destination_program.key
+        );
+        return Err(ProgramError::InvalidAccountData);
+    }
+
+    // verify passed escrow_account derived from delegated_account
```

Comment:
isn't the escrow derived from the authority?

Reply:
Old comment, changed